crates/chunk/examples/chunk_merkle_proving.rs

```diff
@@ -0,0 +1,429 @@
+        .collect();
+    // If the bit size is not a multiple of 253, we need to add the remaining bits
+    if bit_size % F::CAPACITY as usize != 0 {
+        scalar_bit_sizes.push(bit_size % F::CAPACITY as usize)
+    }
+
+    assert_eq!(
+        elements.len(),
+        scalar_bit_sizes.len(),
+        "Got {} elements to reconstruct hash, expected {}",
+        elements.len(),
+        scalar_bit_sizes.len()
+    );
+
+    let mut result: Vec<Boolean> = vec![];
+
+    // For each field element, take the first `bit_size` bits
+    for (i, bit_to_take) in scalar_bit_sizes.iter().enumerate() {
+        let element = elements[i]
+            .to_bits_le(&mut cs.namespace(|| "test"))
+            .unwrap();
+
+        result.extend(element.into_iter().take(*bit_to_take));
+    }
+
+    result
+}
+
+/*****************************************
+ * Circuit
+ *****************************************/
+
+struct MerkleChunkCircuit<F: PrimeField + PrimeFieldBits, C: ChunkStepCircuit<F>, const N: usize> {
+    inner: InnerCircuit<F, C, N>,
+}
+
+impl<F: PrimeField + PrimeFieldBits, C: ChunkStepCircuit<F>, const N: usize>
+    MerkleChunkCircuit<F, C, N>
+{
+    fn new(inputs: &[F], post_processing_step: Option<F>) -> Self {
+        Self {
+            inner: InnerCircuit::new(inputs, post_processing_step).unwrap(),
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+enum ChunkCircuitSet<F: PrimeField + PrimeFieldBits, C: ChunkStepCircuit<F>, const N: usize> {
+    IterStep(FoldStepWrapper<F, C, N>),
+    CheckEquality(EqualityCircuit<F>),
+}
+
+impl<F: PrimeField + PrimeFieldBits, C: ChunkStepCircuit<F>, const N: usize> StepCircuit<F>
+    for ChunkCircuitSet<F, C, N>
+{
+    fn arity(&self) -> usize {
+        match self {
+            Self::IterStep(fold_step) => fold_step.inner.arity(),
+            Self::CheckEquality(equality_circuit) => equality_circuit.arity(),
+        }
+    }
+
+    fn circuit_index(&self) -> usize {
+        match self {
+            Self::IterStep(fold_step) => *fold_step.inner.step_nbr(),
+            Self::CheckEquality(equality_circuit) => equality_circuit.circuit_index(),
+        }
+    }
+
+    fn synthesize<CS: ConstraintSystem<F>>(
+        &self,
+        cs: &mut CS,
+        pc: Option<&AllocatedNum<F>>,
+        z: &[AllocatedNum<F>],
+    ) -> Result<(Option<AllocatedNum<F>>, Vec<AllocatedNum<F>>), SynthesisError> {
+        match self {
+            Self::IterStep(fold_step) => fold_step.synthesize(cs, pc, z),
+            Self::CheckEquality(equality_circuit) => equality_circuit.synthesize(cs, pc, z),
+        }
+    }
+}
+
+impl<E1: CurveCycleEquipped, C: ChunkStepCircuit<E1::Scalar>, const N: usize> NonUniformCircuit<E1>
+    for MerkleChunkCircuit<E1::Scalar, C, N>
+{
+    type C1 = ChunkCircuitSet<E1::Scalar, C, N>;
+    type C2 = TrivialSecondaryCircuit<<Dual<E1> as Engine>::Scalar>;
+
+    fn num_circuits(&self) -> usize {
+        self.inner.num_fold_steps() + 1
+    }
```

Comment:
This seems quite wrong, but maybe I misunderstand the purpose of the chunk gadget. This seems to imply that a computation that is folded 100 times will have 101 distinct circuit shapes. If that were true there would be little benefit in folding or NIVC. You could just inline all the circuits into one giant flat circuit and use that. I very much suspect this is not what you intend.